File: kotlin-node/karakum/src/jsMain/resources/plugins/convertToPrimitiveSymbolHolder.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isTypeLiteralNode(node)
        && node.members.length === 1
        && ts.isMethodSignature(node.members[0])
        && ts.isComputedPropertyName(node.members[0].name)
        && ts.isPropertyAccessExpression(node.members[0].name.expression)

        && ts.isIdentifier(node.members[0].name.expression.expression)
        && node.members[0].name.expression.expression.text === "Symbol"

        && ts.isIdentifier(node.members[0].name.expression.name)
        && node.members[0].name.expression.name.text === "toPrimitive"
    ) {
        return "js.symbol.ToPrimitiveSymbolHolder"
    }


    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertToPrimitiveSymbolHolder.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.*

val convertToPrimitiveSymbolHolder = createPlugin { node, context, render ->
    nullable {
        ensure(isTypeLiteralNode(node))

        val members = node.members.asArray()
        ensure(members.size == 1)

        val member = members.first()
        ensure(isMethodSignature(member))

        val name = member.name
        ensure(isComputedPropertyName(name))

        val expression = name.expression
        ensure(isPropertyAccessExpression(expression))

        val expressionExpression = expression.expression
        ensure(isIdentifier(expressionExpression))
        ensure(expressionExpression.text == "Symbol")

        val expressionName = expression.name
        ensure(isIdentifier(expressionName))
        ensure(expressionName.text == "toPrimitive")

        "js.symbol.ToPrimitiveSymbolHolder"
    }
}
